### PR TITLE
Enable arbitary size Indices for Octree module

### PR DESCRIPTION
**Pull request by kunaltyagi**
And a bug fix for common module

This brings up the modules ready to:
* common, octree
* ml (since it is mostly independent of common)

Next targets: kdtree, sample_consensus, io

### CODE REVIEW

File: octree/include/pcl/octree/impl/octree_search.hpp [hunk 1]
```diff
@@ -123,7 +120,7 @@ OctreePointCloudSearch<PointT, LeafContainerT, BranchContainerT>::nearestKSearch
 template <typename PointT, typename LeafContainerT, typename BranchContainerT>
 int
 OctreePointCloudSearch<PointT, LeafContainerT, BranchContainerT>::nearestKSearch(
-    int index, int k, std::vector<int>& k_indices, std::vector<float>& k_sqr_distances)
+    int index, int k, Indices& k_indices, std::vector<float>& k_sqr_distances)
```
Comment (SergioRAgostinho) [inline octree/include/pcl/octree/impl/octree_search.hpp:123]: `index_t`

File: octree/include/pcl/octree/impl/octree_search.hpp [hunk 2]
```diff
@@ -191,7 +188,7 @@ int
 OctreePointCloudSearch<PointT, LeafContainerT, BranchContainerT>::radiusSearch(
     int index,
```
Comment (SergioRAgostinho) [inline octree/include/pcl/octree/impl/octree_search.hpp:189]: `index_t`

File: octree/include/pcl/octree/impl/octree_search.hpp [hunk 3]
```diff
@@ -80,10 +80,7 @@ OctreePointCloudSearch<PointT, LeafContainerT, BranchContainerT>::voxelSearch(
 template <typename PointT, typename LeafContainerT, typename BranchContainerT>
 int
 OctreePointCloudSearch<PointT, LeafContainerT, BranchContainerT>::nearestKSearch(
-    const PointT& p_q,
-    int k,
-    std::vector<int>& k_indices,
-    std::vector<float>& k_sqr_distances)
+    const PointT& p_q, int k, Indices& k_indices, std::vector<float>& k_sqr_distances)
```
Comment (SergioRAgostinho) [inline octree/include/pcl/octree/impl/octree_search.hpp:83]: Not actionable: k should likely be made uindex_t in a future pass

File: octree/include/pcl/octree/octree_search.h [hunk 4]
```diff
@@ -100,7 +100,7 @@ class OctreePointCloudSearch
    * \return "true" if leaf node exist; "false" otherwise
    */
   bool
-  voxelSearch(const int index, std::vector<int>& point_idx_data);
+  voxelSearch(const int index, Indices& point_idx_data);
```
Comment (SergioRAgostinho) [inline octree/include/pcl/octree/octree_search.h:103]: index

File: octree/include/pcl/octree/octree_search.h [hunk 5]
```diff
@@ -116,7 +116,7 @@ class OctreePointCloudSearch
   nearestKSearch(const PointCloud& cloud,
                  int index,
```
Comment (SergioRAgostinho) [inline octree/include/pcl/octree/octree_search.h:117]: ping

File: octree/include/pcl/octree/octree_search.h [hunk 6]
```diff
@@ -203,7 +203,7 @@ class OctreePointCloudSearch
   radiusSearch(const PointCloud& cloud,
                int index,
```
Comment (SergioRAgostinho) [inline octree/include/pcl/octree/octree_search.h:204]: ping

File: octree/include/pcl/octree/octree_search.h [hunk 7]
```diff
@@ -240,7 +240,7 @@ class OctreePointCloudSearch
   int
   radiusSearch(int index,
```
Comment (SergioRAgostinho) [inline octree/include/pcl/octree/octree_search.h:241]: ping

**Review by SergioRAgostinho — CHANGES_REQUESTED** [review]
It doesn't work to simply look at method's interfaces. There's some rotten apples inside these methods which will still break functionality. 
https://github.com/PointCloudLibrary/pcl/blob/c52674ec202de8d36755fcdc0230565248d4b7d7/octree/include/pcl/octree/impl/octree_search.hpp#L107

I would rather take a more careful approach to this migration.

**Comment by kunaltyagi** [comment]
Marking draft. Will return with vengeance (every line checked... 😢)

File: octree/include/pcl/octree/impl/octree_search.hpp (hunk 3 above)
Comment (kunaltyagi) [reply octree/include/pcl/octree/impl/octree_search.hpp:83]: `k` is currently signed. Should it be make `index_t` instead?

**Comment by kunaltyagi** [comment]
Several use of `int` left in octree_pointcloud.hpp
PTAL

File: kdtree/include/pcl/kdtree/kdtree_flann.h [hunk 8]
```diff
@@ -178,7 +178,7 @@ namespace pcl
         * \exception asserts in debug mode if the index is not between 0 and the maximum number of points
         */
       int 
-      radiusSearch (const PointT &point, double radius, std::vector<int> &k_indices,
+      radiusSearch (const PointT &point, double radius, Indices &k_indices,
                     std::vector<float> &k_sqr_distances, unsigned int max_nn = 0) const override;
```
Comment (larshg) [inline kdtree/include/pcl/kdtree/kdtree_flann.h:182]: should max_nn be index_t as well - since its the max numbers of indexes?

File: octree/include/pcl/octree/impl/octree_pointcloud.hpp [hunk 9]
```diff
@@ -247,7 +248,7 @@ template <typename PointT,
           typename LeafContainerT,
           typename BranchContainerT,
           typename OctreeT>
-int
+pcl::uindex_t
```
Comment (larshg) [inline octree/include/pcl/octree/impl/octree_pointcloud.hpp:251]: index_t here as in line 268?

File: octree/include/pcl/octree/impl/octree_pointcloud.hpp [hunk 10]
```diff
@@ -962,13 +952,13 @@ template <typename PointT,
...
```
Comment (larshg) [inline octree/include/pcl/octree/impl/octree_pointcloud.hpp:955]: Similar as before, with these get*Centers*. index_t or all uindex_t?

File: octree/include/pcl/octree/impl/octree_pointcloud_adjacency.hpp [hunk 11]
```diff
@@ -295,13 +289,13 @@ pcl::octree::OctreePointCloudAdjacency<PointT, LeafContainerT, BranchContainerT>
   direction.normalize();
   float precision = 1.0f;
   const float step_size = static_cast<const float>(resolution_) * precision;
-  const int nsteps = std::max(1, static_cast<int>(norm / step_size));
+  const auto nsteps = std::max<uindex_t>(1, norm / step_size);
```
Comment (larshg) [inline octree/include/pcl/octree/impl/octree_pointcloud_adjacency.hpp:292]: at line 248 you selected std::size_t?

File: octree/include/pcl/octree/impl/octree_pointcloud_adjacency.hpp [hunk 12]
```diff
@@ -295,13 +289,13 @@ pcl::octree::OctreePointCloudAdjacency<PointT, LeafContainerT, BranchContainerT>
...
 
   OctreeKey prev_key = key;
   // Walk along the line segment with small steps.
   Eigen::Vector3f p = leaf_centroid;
   PointT octree_p;
-  for (int i = 0; i < nsteps; ++i) {
+  for (uindex_t i = 0; i < nsteps; ++i) {
```
Comment (larshg) [inline octree/include/pcl/octree/impl/octree_pointcloud_adjacency.hpp:298]: connected with 292 - std::size_t?

File: octree/include/pcl/octree/impl/octree_search.hpp [hunk 13]
```diff
@@ -107,23 +107,23 @@ OctreePointCloudSearch<PointT, LeafContainerT, BranchContainerT>::nearestKSearch
...
       p_q, k, this->root_node_, key, 1, smallest_dist, point_candidates);
 
-  unsigned int result_count = static_cast<unsigned int>(point_candidates.size());
+  index_t result_count = static_cast<index_t>(point_candidates.size());
 
   k_indices.resize(result_count);
   k_sqr_distances.resize(result_count);
 
-  for (unsigned int i = 0; i < result_count; ++i) {
+  for (index_t i = 0; i < result_count; ++i) {
```
Comment (larshg) [inline octree/include/pcl/octree/impl/octree_search.hpp:115]: uindex_t?

File: octree/include/pcl/octree/impl/octree_search.hpp [hunk 14]
```diff
@@ -107,23 +107,23 @@ OctreePointCloudSearch<PointT, LeafContainerT, BranchContainerT>::nearestKSearch
   getKNearestNeighborRecursive(
       p_q, k, this->root_node_, key, 1, smallest_dist, point_candidates);
 
-  unsigned int result_count = static_cast<unsigned int>(point_candidates.size());
+  index_t result_count = static_cast<index_t>(point_candidates.size());
```
Comment (larshg) [inline octree/include/pcl/octree/impl/octree_search.hpp:110]: uindex_t?

File: octree/include/pcl/octree/octree_pointcloud_changedetector.h [hunk 15]
```diff
@@ -91,15 +91,15 @@ class OctreePointCloudChangeDetector
    * \return number of point indices
    */
   std::size_t
-  getPointIndicesFromNewVoxels(std::vector<int>& indicesVector_arg,
-                               const int minPointsPerLeaf_arg = 0)
+  getPointIndicesFromNewVoxels(Indices& indicesVector_arg,
+                               const index_t minPointsPerLeaf_arg = 0)
```
Comment (larshg) [inline octree/include/pcl/octree/octree_pointcloud_changedetector.h:95]: unindex_t - can't have a negative min count?

File: octree/include/pcl/octree/octree_pointcloud_adjacency_container.h [hunk 16]
```diff
@@ -117,10 +117,10 @@ class OctreePointCloudAdjacencyContainer : public OctreeContainerBase {
   /** \brief  virtual method to get size of container
    * \return number of points added to leaf node container.
    */
-  std::size_t
+  index_t
```
Comment (larshg) [inline octree/include/pcl/octree/octree_pointcloud_adjacency_container.h:120]: uindex_t? and remove unnecssary cast as num_points_ is uindex_t?

Comment (larshg) [reply octree/include/pcl/octree/octree_pointcloud_adjacency_container.h:120]: getPointCounter and getSize returns the same variable value? Deprecate getPointCounter? (maybe another pr)

File: octree/include/pcl/octree/octree_pointcloud_adjacency.h [hunk 17]
```diff
@@ -197,7 +197,7 @@ class OctreePointCloudAdjacency
    * \note This virtual implementation allows the use of a transform function to compute
    * keys. */
   void
-  addPointIdx(const int point_idx_arg) override;
+  addPointIdx(index_t point_idx_arg) override;
```
Comment (larshg) [inline octree/include/pcl/octree/octree_pointcloud_adjacency.h:200]: uindex_t as you can use negative numbers as index to a data structure?

File: octree/include/pcl/octree/octree_pointcloud.h [hunk 18]
```diff
@@ -440,15 +440,15 @@ class OctreePointCloud : public OctreeT {
...
+                 uindex_t depth_mask);
 
   /** \brief Get point at index from input pointcloud dataset
    * \param[in] index_arg index representing the point in the dataset given by \a
    * setInputCloud
    * \return PointT from input pointcloud dataset
    */
   const PointT&
-  getPointByIndex(const unsigned int index_arg) const;
+  getPointByIndex(index_t index_arg) const;
```
Comment (larshg) [inline octree/include/pcl/octree/octree_pointcloud.h:451]: uindex_t?

File: octree/include/pcl/octree/octree_pointcloud.h [hunk 19]
```diff
@@ -428,7 +428,7 @@ class OctreePointCloud : public OctreeT {
    * \a setInputCloud to be added
    */
   virtual void
-  addPointIdx(const int point_idx_arg);
+  addPointIdx(index_t point_idx_arg);
```
Comment (larshg) [inline octree/include/pcl/octree/octree_pointcloud.h:431]: uindex_t?

File: octree/include/pcl/octree/octree_pointcloud.h [hunk 20]
```diff
@@ -295,7 +295,7 @@ class OctreePointCloud : public OctreeT {
    *  \param[in] point_idx_arg index of point addressing the voxel to be deleted.
    */
   void
-  deleteVoxelAtPoint(const int& point_idx_arg);
+  deleteVoxelAtPoint(const index_t& point_idx_arg);
```
Comment (larshg) [inline octree/include/pcl/octree/octree_pointcloud.h:298]: uindex_t?

File: octree/include/pcl/octree/octree_pointcloud.h [hunk 21]
```diff
@@ -258,14 +258,14 @@ class OctreePointCloud : public OctreeT {
    * \return "true" if voxel exist; "false" otherwise
    */
   bool
-  isVoxelOccupiedAtPoint(const int& point_idx_arg) const;
+  isVoxelOccupiedAtPoint(const index_t& point_idx_arg) const;
```
Comment (larshg) [inline octree/include/pcl/octree/octree_pointcloud.h:261]:
```suggestion
  isVoxelOccupiedAtPoint(const uindex_t& point_idx_arg) const;
```

File: octree/include/pcl/octree/octree_pointcloud.h [hunk 22]
```diff
@@ -200,7 +200,7 @@ class OctreePointCloud : public OctreeT {
    * setInputCloud)
    */
   void
-  addPointFromCloud(const int point_idx_arg, IndicesPtr indices_arg);
+  addPointFromCloud(index_t point_idx_arg, IndicesPtr indices_arg);
```
Comment (larshg) [inline octree/include/pcl/octree/octree_pointcloud.h:203]: uindex_t

File: octree/include/pcl/octree/octree_container.h [hunk 23]
```diff
@@ -314,10 +314,10 @@ class OctreeContainerPointIndices : public OctreeContainerBase {
   /** \brief Get size of container (number of indices)
    * \return number of point indices in container.
    */
-  std::size_t
+  index_t
```
Comment (larshg) [inline octree/include/pcl/octree/octree_container.h:317]: uindex_t - can't have a negative size?

File: octree/include/pcl/octree/octree_container.h [hunk 24]
```diff
@@ -125,7 +127,7 @@ class OctreeContainerEmpty : public OctreeContainerBase {
   /** \brief Abstract get size of container (number of DataT objects)
    * \return number of DataT elements in leaf node container.
    */
-  std::size_t
+  index_t
```
Comment (larshg) [inline octree/include/pcl/octree/octree_container.h:130]: uindex_t

File: octree/include/pcl/octree/octree_container.h [hunk 25]
```diff
@@ -74,7 +76,7 @@ class OctreeContainerBase {
   /** \brief Pure abstract method to get size of container (number of indices)
    * \return number of points/indices stored in leaf node container.
    */
-  virtual std::size_t
+  virtual index_t
```
Comment (larshg) [inline octree/include/pcl/octree/octree_container.h:79]: uindex_t

**Comment by larshg** [comment]
I think we should aim to use uindex_t as much as possible, as index as a negative number doesn't make sense.
There are though following exceptions (due to bad implementation in first place):

- Searching for something and returning -1 as not found. which would require index_t for now. Until refactored.
- other cases?

File: kdtree/include/pcl/kdtree/kdtree_flann.h (hunk 8 above)
Comment (kunaltyagi) [reply kdtree/include/pcl/kdtree/kdtree_flann.h:182]: I'm touching other modules enough to just compile in 64 bit mode on Linux

File: octree/include/pcl/octree/impl/octree_pointcloud.hpp (hunk 10 above)
Comment (kunaltyagi) [reply octree/include/pcl/octree/impl/octree_pointcloud.hpp:955]: I had missed one (all uindex_t), since I started with index_t, then thought, hey, these are never negative

File: octree/include/pcl/octree/impl/octree_pointcloud_adjacency.hpp (hunk 11 above)
Comment (kunaltyagi) [reply octree/include/pcl/octree/impl/octree_pointcloud_adjacency.hpp]: Good catch. It should be size

**Review by kunaltyagi — COMMENTED** [review]
indices can't be negative, but they are of type `index_t`. `uindex_t` is a promise from the API that the value is never negative, which is not to say that this is a valid index.

Should all indices be unsigned? Not yet (also because it'd be a huge risky API break), but in due time yes. Right now, lets focussing only on getting PCL compiled for 32 and 64 bits, signed and unsigned. Once that's done, we can shift focus to compare warnings, and after that we can make index_t and uindex_t the same.

**Comment by kunaltyagi** [comment]
> I think we should aim to use uindex_t as much as possible, as index as a negative number doesn't make sense.

Didn't read until after I wrote the comment above. I completely agree in theory, but there needs to be a roadmap from `int` -> `unsigned whatever-size`. That's why the separation between index_t and uindex_t needs to exist practically. index_t is 100% compatible with int with the right options. Once we have everything compiling without warnings with PCL_INDEX_SIGNED as false and PCL_INDEX_SIZE as 32 and 64, we can move all indices to unsigned land. Hope this makes more sense.

File: octree/include/pcl/octree/octree_pointcloud_adjacency_container.h (hunk 16 above)
Comment (kunaltyagi) [reply octree/include/pcl/octree/octree_pointcloud_adjacency_container.h]: OctreePointCloudDensity has a getPointCounter but not getSize. I agree in principle (someone should do it)

**Comment by mvieth** [comment]
@kunaltyagi Do you maybe want to check why the tests are failing and resolve the conflicts? Then we could hopefully merge this soon.

**Comment by kunaltyagi** [comment]
@mvieth Thanks a lot for pinging me. I've been so busy, it's been hard to keep track of what's the priority here. :)

Just rebased to current master and will get back to this in this coming week.

File: octree/include/pcl/octree/octree_pointcloud.h (hunk 19 above)
Comment (larshg) [reply octree/include/pcl/octree/octree_pointcloud.h:431]: Should const stay? (multiple occurencies)

**Comment by kunaltyagi** [comment]
@larshg I'll take a look at your existing comments on the weekend.

File: octree/include/pcl/octree/octree_pointcloud.h (hunk 18 above)
Comment (kunaltyagi) [reply octree/include/pcl/octree/octree_pointcloud.h]: Please see commit with message "Making incoming index unsigned"

File: octree/include/pcl/octree/octree_pointcloud.h (hunk 19 above)
Comment (kunaltyagi) [reply octree/include/pcl/octree/octree_pointcloud.h]: `const` on a by-value parameter doesn't have any influence on the API. If only makes sense in a definition, not declaration (since it doesn't impose any constraints)

File: octree/include/pcl/octree/octree_pointcloud.h (hunk 20 above)
Comment (kunaltyagi) [reply octree/include/pcl/octree/octree_pointcloud.h:298]: This is a reference, and all of our indices are `index_t`. Having an unsigned reference to a signed value muddies things up.

I've converted all the by-value to unsigned, but the reference ones can't be helped

**Review by larshg — APPROVED** [review]

File: octree/include/pcl/octree/impl/octree_iterator.hpp [hunk 26]
```diff
@@ -258,14 +257,13 @@ OctreeBreadthFirstIterator<OctreeT>::operator++()
 
 //////////////////////////////////////////////////////////////////////////////////////////////
 template <typename OctreeT>
-OctreeFixedDepthIterator<OctreeT>::OctreeFixedDepthIterator()
-: OctreeBreadthFirstIterator<OctreeT>(0u), fixed_depth_(0u)
```
Comment (mvieth) [inline octree/include/pcl/octree/impl/octree_iterator.hpp:262]: Why did you change this line?

**Review by mvieth — COMMENTED** [review]
I think it is sometimes ambiguous whether `index_t` or `uindex_t` should be used, but I generally agree with your choices.

File: octree/include/pcl/octree/impl/octree_iterator.hpp (hunk 26 above)
Comment (kunaltyagi) [reply octree/include/pcl/octree/impl/octree_iterator.hpp]: Accidentally? I don't recall making this change. I might have deleted something by accident and the formatter updated the line.

File: io/include/pcl/compression/color_coding.h [hunk 27]
```diff
@@ -155,7 +155,7 @@ class ColorCoding
    * \param inputCloud_arg input point cloud
    * */
   void
-  encodeAverageOfPoints (const typename std::vector<int>& indexVector_arg, unsigned char rgba_offset_arg, PointCloudConstPtr inputCloud_arg)
+  encodeAverageOfPoints (const Indices& indexVector_arg, unsigned char rgba_offset_arg, PointCloudConstPtr inputCloud_arg)
```
[the remark this answers is not in the document]
Comment (kunaltyagi) [reply io/include/pcl/compression/color_coding.h:158]: I noticed a few changes that needed to be propagated to ensure correctness. PTAL

File: io/include/pcl/compression/color_coding.h [hunk 28]
```diff
@@ -281,12 +278,12 @@ class ColorCoding
     * \param rgba_offset_arg offset to color information
     */
   void
-  decodePoints (PointCloudPtr outputCloud_arg, std::size_t beginIdx_arg, std::size_t endIdx_arg, unsigned char rgba_offset_arg)
+  decodePoints (PointCloudPtr outputCloud_arg, index_t beginIdx_arg, index_t endIdx_arg, unsigned char rgba_offset_arg)
```
Comment (larshg) [inline io/include/pcl/compression/color_coding.h:281]: Size_t is unsigned right? So could be uindex_t?

Comment (kunaltyagi) [reply io/include/pcl/compression/color_coding.h]: My bad

**Review by mvieth — APPROVED** [review]
LGTM, assuming you fix the failing CI check.

**Comment by mvieth** [comment]
@kunaltyagi Merge?

**Comment by kunaltyagi** [comment]
Rearranged the comments a bit for an easy(er) merge
